source/common/jwt_authn/jwks.cc

```diff
@@ -0,0 +1,249 @@
+#include "common/jwt_authn/jwks.h"
+
+#include "common/common/assert.h"
+#include "common/common/base64.h"
+#include "common/json/json_loader.h"
+#include "common/jwt_authn/utils.h"
+
+#include "openssl/bn.h"
+#include "openssl/ecdsa.h"
+#include "openssl/evp.h"
+#include "openssl/rsa.h"
+#include "openssl/sha.h"
+
+namespace Envoy {
+namespace JwtAuthn {
+
+namespace {
+
+/** Class to create EVP_PKEY object from string of public key, formatted in PEM or JWKs.
+ * If it failed, status_ holds the failure reason.
+ *
+ * Usage example:
+ * EvpPkeyGetter e;
+ * bssl::UniquePtr<EVP_PKEY> pkey = e.createEvpPkeyFromStr(pem_formatted_public_key);
+ * (You can use createEvpPkeyFromJwkRSA() or createEcKeyFromJwkEC() for JWKs)
+ */
+class EvpPkeyGetter : public WithStatus {
+public:
+  // Create EVP_PKEY from PEM string
+  bssl::UniquePtr<EVP_PKEY> createEvpPkeyFromStr(const std::string& pkey_pem) {
+    std::string pkey_der = Base64::decode(pkey_pem);
+    if (pkey_der == "") {
+      updateStatus(Status::JwksPemBadBase64);
+      return nullptr;
+    }
+    auto rsa =
+        bssl::UniquePtr<RSA>(RSA_public_key_from_bytes(castToUChar(pkey_der), pkey_der.length()));
+    if (!rsa) {
+      updateStatus(Status::JwksPemParseError);
+    }
+    return createEvpPkeyFromRsa(rsa.get());
+  }
+
+  bssl::UniquePtr<EVP_PKEY> createEvpPkeyFromJwkRSA(const std::string& n, const std::string& e) {
+    return createEvpPkeyFromRsa(createRsaFromJwk(n, e).get());
+  }
+
+  bssl::UniquePtr<EC_KEY> createEcKeyFromJwkEC(const std::string& x, const std::string& y) {
+    bssl::UniquePtr<EC_KEY> ec_key(EC_KEY_new_by_curve_name(NID_X9_62_prime256v1));
+    if (!ec_key) {
+      updateStatus(Status::JwksEcCreateKeyFail);
+      return nullptr;
+    }
+    bssl::UniquePtr<BIGNUM> bn_x = createBigNumFromBase64UrlString(x);
+    bssl::UniquePtr<BIGNUM> bn_y = createBigNumFromBase64UrlString(y);
+    if (!bn_x || !bn_y) {
+      // EC public key field is missing or has parse error.
+      updateStatus(Status::JwksEcParseError);
+      return nullptr;
+    }
+
+    if (EC_KEY_set_public_key_affine_coordinates(ec_key.get(), bn_x.get(), bn_y.get()) == 0) {
+      updateStatus(Status::JwksEcParseError);
+      return nullptr;
+    }
+    return ec_key;
+  }
+
+private:
+  bssl::UniquePtr<EVP_PKEY> createEvpPkeyFromRsa(RSA* rsa) {
+    if (!rsa) {
+      return nullptr;
+    }
+    bssl::UniquePtr<EVP_PKEY> key(EVP_PKEY_new());
+    EVP_PKEY_set1_RSA(key.get(), rsa);
+    return key;
+  }
+
+  bssl::UniquePtr<BIGNUM> createBigNumFromBase64UrlString(const std::string& s) {
+    std::string s_decoded = decodeBase64Url(s);
+    if (s_decoded == "") {
+      return nullptr;
+    }
+    return bssl::UniquePtr<BIGNUM>(BN_bin2bn(castToUChar(s_decoded), s_decoded.length(), NULL));
+  };
+
+  bssl::UniquePtr<RSA> createRsaFromJwk(const std::string& n, const std::string& e) {
+    bssl::UniquePtr<RSA> rsa(RSA_new());
+    // It crash if RSA object couldn't be created.
+    ASSERT(rsa);
```

Comment:
I don't think you need this ASSERT, it will crash right next line if it is nullptr

Reply:
Removed